sdks/java/harness/src/main/java/org/apache/beam/fn/harness/BeamFnDataWriteRunner.java

```diff
@@ -66,93 +61,57 @@
   }
 
   /** A factory for {@link BeamFnDataWriteRunner}s. */
-  static class Factory<InputT> implements PTransformRunnerFactory<BeamFnDataWriteRunner<InputT>> {
+  static class Factory<InputT> implements PTransformRunnerFactory<BeamFnDataWriteRunner> {
 
     @Override
-    public BeamFnDataWriteRunner<InputT> createRunnerForPTransform(Context context)
-        throws IOException {
-
-      BeamFnDataWriteRunner<InputT> runner =
-          new BeamFnDataWriteRunner<>(
-              context.getBundleCacheSupplier(),
-              context.getPTransformId(),
-              context.getPTransform(),
-              context.getProcessBundleInstructionIdSupplier(),
-              context.getCoders(),
-              context.getBeamFnDataClient(),
-              context.getBeamFnStateClient());
-      context.addStartBundleFunction(runner::registerForOutput);
+    public BeamFnDataWriteRunner createRunnerForPTransform(Context context) throws IOException {
+
+      RemoteGrpcPort port = RemoteGrpcPortWrite.fromPTransform(context.getPTransform()).getPort();
+      RehydratedComponents components =
+          RehydratedComponents.forComponents(
+              Components.newBuilder().putAllCoders(context.getCoders()).build());
+      Coder<WindowedValue<InputT>> coder =
+          (Coder<WindowedValue<InputT>>)
+              CoderTranslation.fromProto(
+                  context.getCoders().get(port.getCoderId()),
+                  components,
+                  new StateBackedIterableTranslationContext() {
+                    @Override
+                    public Supplier<Cache<?, ?>> getCache() {
+                      return context.getBundleCacheSupplier();
+                    }
+
+                    @Override
+                    public BeamFnStateClient getStateClient() {
+                      return context.getBeamFnStateClient();
+                    }
+
+                    @Override
+                    public Supplier<String> getCurrentInstructionId() {
+                      return context.getProcessBundleInstructionIdSupplier();
+                    }
+                  });
+      BeamFnDataWriteRunner<InputT> runner = new BeamFnDataWriteRunner<>();
+      context.addStartBundleFunction(
```

Comment:
I was thinking that `context#addOutgoingDataEndpoint` would happen now and not within start bundle. This would also allow us to pass the FnDataReceiver to the `context#addPCollectionConsumer` directly instead of registering `runner::consume`.